# 31st March, 2021

## TODOs 

1. Writing on PID
2. fcFDR applied to IMDs


## What I've done

### Writing on PID

I had a bit of a rough week personally (sleep, lockdown) and I didn't get my writing done as a result.

### fcFDR applied to PID and the IMDs

#### Bug from last week

A bug held me up last week but I talked with Anna and we've since resolved this. It turned out that `spatstat` had since been split up into several packages and we needed to import the relevant function (`owin`) from `spatstat.geom` instead.

#### Comparing fcFDR with and without matching

In short, there appears to be no easily discernible differences between the results produced with and without matching, although I still think matching is the judicious thing to do. In addition, downsampling the independent SNPs cuts down on the running time, although `fcFDR` is very quick in comparison to `cFDR`, anyway.

Manhattans in case you would like to look, but I can't see the difference. Note that I carried over the T1D minor allele frequencies here. 

[Comparing fcFDR, matched and unmatched](/entries/310321/matched_vs_unmatched.html)

Running times:

| Data set | Elapsed (unmatched) | Elapsed (matched) |
|----------|---------------------|-------------------|
| t1d      |             0:34:19 |           0:24:44 |
| sle      |             0:34:46 |           0:24:42 |
| ra       |             0:23:48 |           0:18:24 |
| aster    |             0:25:33 |           0:18:35 |
| cd       |             0:33:37 |           0:24:42 |
| uc       |             0:35:53 |           0:35:39 |
| asthma   |             0:49:37 |           0:40:45 |
| cad      |             0:30:38 |           0:22:54 |

#### Comparing fcFDR with cFDR on IMDs

The fcFDR results were worrying for want of genome-wide significant SNPs: of seven IMDs, only one (asthma/allergy/eczema) yielded a borderline significant SNP on chromosome 4. 

[Comparing fcFDR and cFDR on IMDs](/entries/310321/cfdr_vs_fcfdr.html)

Why is this?

#### What do exploratory plots suggest about enrichment in the IMDs?

The PID data set consists of GWAS summary statistics of antibody-deficient PID of SNPs of intermediate and common frequency MAF, > 0.005 (I think, anyway, it's not made entirely clear in the EBI GWAS Catalog nor the paper itself which data set was uploaded). There was only one genome-wide significant tower outside of the MHC, at chromosome 17, so it was not the most signal-rich data set. But then again, there was a plethora of borderline/suggesitve signals and are we not using cFDR methods to 'promote' such signals to significance in the first place?

![](/images/030321/pid.png)

The stratified Q-Q plots we've looked at it in the past provide encouraging signs of enrichment among the IMD data sets. However, I generated correlation plots using Anna's code from the `fcfdr` package and these seem to show no sign of enrichment. Nonetheless, `cfdr` is clearly picking up something in the same data sets. 

#### Are we losing SNPs?

I've often returned to the issue of how performing inner joins on data sets leads to our losing some SNPs, whether we are joining two sets of summary statistics, summary statistics and MAF values (from T1D), or summary statistics and LDAK weights (in practice this amounts to joining summary statistic SNPs and those in the 1000G data). In the case of the `fcfdr`, we've performed all three such joins, whereas the `cfdr` data sets have been through only the first of these joins. It shows: the `cfdr` data sets may have one million more SNPs than the `fcfdr` data sets (in the case of the `aster` data). How do the Manhattan plots differ when we downsample the `cfdr` results to have the same SNPs as in the `fcfdr` results? 

It's not a fair comparison as the `cfdr` procedure still gets the benefit of those SNPs in the actual procedure. Nonetheless, `cfdr` is still more powerful when looking only at these shared SNPs.

[Comparing fcFDR and cFDR on IMDs at shared SNPs](/entries/310321/cfdr_vs_fcfdr_sharedOnly.html)

So we are missing some SNPs in the `fcfdr` data sets. Are they the important ones? I looked at the top 20 SNPs per chromosome in the PID GWAS and whether they were retained in the `fcfdr` data sets. 

| Data set | No. of missing SNPs from top 20 per chrom. | % missing |
|----------|--------------------------------------------|-----------|
| aster    |                                        182 | 41%       |
| asthma   |                                        214 | 49%       |
| cd       |                                        182 | 41%       |
| ra       |                                        263 | 60%       |
| sle      |                                        191 | 43%       |
| t1d      |                                        168 | 31%       |
| uc       |                                        182 | 41%       |

(NB: I've found the RA data set to be an outlier in general when it comes to joining.)

These numbers don't really tell us if we are losing *signals*, though. We might tolerate dropping a few SNPs from a tower provided it's still visible in the Manhattan. I highlighted the top 20 PID SNPs in the results to see if signals were retained in `fcfdr` results.

[Highlighting top SNPs in fcFDR results](/entries/310321/missing_top20.html)

This is admittedly not the best way to visualise the results, but it seems apparent we are missing some signals, particularly those on chromosome 17 that accounted for the only extra-MHC GWS signal in the original GWAS.

Unfortunately, there's not really a way around this. Even if we do not recycle the T1D MAF values, the only alternative is to compute them from the 1000G data, from which we already lose SNPs when joining the data to the LDAK weights.

#### What about non-PID data? T1D on SLE 

How does `fcfdr` perform on a pair of more common diseases, T1D and SLE? These are known to share genetic architecture and, in being more common, should give us more 'conventional' GWAS results. 

Comparing the `fcfdr` results to those from `cfdr` and not restricting to common SNPs:

![](/images/310321/t1d_on_sle_cfdr_vs_fcfdr.png)

I do not like these results as you could almost believe I just copied the original T1D p-values to the fcFDR track in lieu of the v-values (I even checked that I had not made exactly that mistake). `cfdr` is doing something: see the peak at chr19, for example. 

If we look at the `fcfdr` alone with the T1D p-values and the SLE p-values (that is, the principal and auxiliary traits), I'm no happier:

![](/images/310321/t1d_on_sle.png)

#### The upshot

After putting all of this together, I'm going to scour my code to see if the script I've written to run `fcfdr` is doing something wrong. I can't shift the feeling I'm doing something incorrectly somewhere in my workflow.
