# 7th April, 2021

## TODOs 

1. Try supplying as auxiliary Z-values obtained with `qnorm(p/2)` rather than `-qnorm(p/2)`
2. How many SNPs are we losing in PID joins? (Recall in particular T1D-PID join loses none due to MAF join)
3. Use hg38 1000G data for LDAK, see if it improves number of SNPs retained

## What I've done

### Trying `qnorm(p/2)` instead of `-qnorm(p/2)`

Chris suggested that I flip the sign on the auxiliary Z-values as `-qnorm(p/2)` produces Z-values which are negatively correlated with the underlying auxiliary p-values. I had thought that, as `fcfdr::flexible_cfdr` transforms the principal p-values in this way, that it would be ok to do the same for p-values supplied as the auxiliary covariate. However, this approach would produce a negative correlation with the principal p-values in a setting of shared genetic architecture where the principal and auxiliary p-values are positively correlated.

Flipping the sign in front of `qnorm` seemed to work and fcFDR boosted one or more signals to genome-wide significance in five of the seven IMDs. Recall that we had previously seen no newly significant loci when using fcFDR (at most one borderline SNP from conditioning on the allergy GWAS).

The 'allergy phenotype' (allergic rhinitis, eczema, and asthma) GWAS provided the greatest boost, with four newly significant loci. The T1D data yielded four, too, but the allergy data additionally shifted a few suggestive SNPs to a point just short of significance.

![](/images/070421/un_flipped_manhattans/aster_flipped.png)

[Manhattans comparing the results obtained with each set of Z-values](/entries/070421/un_flipped_manhattans.html)

[Manhattans with PID, IMD auxiliary values, and fcFDR results](/entries/070421/aux_vis_manhattans.html)

### Comparing these new results to the old cFDR results

How do these new results compare to the cFDR results? Note that in the following plots, the PID track is drawn using the principal p-values *retained* in the fcFDR data set in each case. Notably, we lose a group of SNPs reaching genome-wide significance

[Manhattans comparing fcFDR and cFDR results](/entries/070421/fcfdr_vs_cfdr.html)

[Manhattans comparing fcFDR and cFDR results, shared SNPs only](/images/070421/fcfdr_vs_cfdr_shared.html)

The sign of Spearman's rank correlation between the principal and auxiliary values (in the independent subset of SNPs) does not seem to predict the no. of newly significant loci or the degree of boosting. Seems a bit daft to think it would do so in any precise way, of course, but I thought I would check.

 | Data set | Spearman's corr. | Pearson's corr. | # newly significant signals (see above) |
 |----------|------------------|-----------------|-----------------------------------------|
 | t1d      |           0.0028 |          0.0039 |                                       4 |
 | asthma   |          -0.0006 |          0.0001 |                                       0 |
 | uc       |           0.0018 |          0.0018 |                                       1 |
 | cd       |           0.0008 |          0.0001 |                                       0 |
 | sle      |           0.0017 |          0.0041 |                                       1 |
 | aster    |          -0.0008 |         -0.0001 |                                       4 |
 | cad      |          -0.0018 |         -0.0009 |                                       0 |
 | ra       |           0.0021 |          0.0037 |                                       1 |
