# 7th April, 2021

## TODOs 

1. Try supplying as auxiliary Z-values obtained with `qnorm(p/2)` rather than `-qnorm(p/2)`
2. How many SNPs are we losing in PID joins? (Recall in particular T1D-PID join loses none due to MAF join)
3. Use hg38 1000G data for LDAK, see if it improves number of SNPs retained

## What I've done

### Trying `qnorm(p/2)` instead of `-qnorm(p/2)`

Chris suggested that I flip the sign on the auxiliary Z-values as `-qnorm(p/2)` produces Z-values which are negatively correlated with the underlying auxiliary p-values. I had thought that, as `fcfdr::flexible_cfdr` transforms the principal p-values in this way, that it would be ok to do the same for p-values supplied as the auxiliary covariate. However, this approach would produce a negative correlation with the principal p-values in a setting of shared genetic architecture where the principal and auxiliary p-values are positively correlated.

Flipping the sign in front of `qnorm` seemed to work and fcFDR boosted one or more signals to genome-wide significance in five of the seven IMDs. Recall that we had previously seen no newly significant loci when using fcFDR (at most one borderline SNP from conditioning on the allergy GWAS).

The 'allergy phenotype' (allergic rhinitis, eczema, and asthma) GWAS provided the greatest boost, with four newly significant loci. The T1D data yielded four, too, but the allergy data additionally shifted a few suggestive SNPs to a point just short of significance.

![](/images/070421/un_flipped_manhattans/aster_flipped.png)

[Manhattans comparing the results obtained with each set of Z-values](/entries/070421/un_flipped_manhattans.html)

[Manhattans with PID, IMD auxiliary values, and fcFDR results](/entries/070421/aux_vis_manhattans.html)

### Comparing these new results to the old cFDR results

How do these new results compare to the cFDR results? Note that in the following plots, the PID track is drawn using the principal p-values *retained* in the fcFDR data set in each case, i.e. after the auxiliary-PID, GWAS-LDAK, and GWAS-MAF joins. It is worth pointing out that we lose a trio of highly significant SNPs on chromosome 17 in the fcFDR joins and most of the cFDR ones, too, although one of those SNPs is retained in several of the cFDR data sets.

I produced Manhattans which compare the fcFDR and cFDR results. The second link below depicts these Manhattans after the cFDR results have been downsampled to retain only those SNPs present in the fcFDR set (although the cFDR procedure has had the benefit of those SNPs in the fitting process).

[Manhattans comparing fcFDR and cFDR results](/entries/070421/fcfdr_vs_cfdr.html)

[Manhattans comparing fcFDR and cFDR results, shared SNPs only](/entries/070421/fcfdr_vs_cfdr_shared.html)

The sign of Spearman's rank correlation between the principal and auxiliary values (in the independent subset of SNPs) does not seem to predict the no. of newly significant loci or the degree of boosting. Seems a bit daft to think it would do so in any precise way, of course, but I thought I would check.

 | Data set | Spearman's corr. | Pearson's corr. | # newly significant signals (see above) |
 |----------|------------------|-----------------|-----------------------------------------|
 | T1D   |           0.0028 |          0.0039 |                                       4 |
 | Asthma |          -0.0006 |          0.0001 |                                       0 |
 | UC     |           0.0018 |          0.0018 |                                       1 |
 | CD     |           0.0008 |          0.0001 |                                       0 |
 | SLE   |           0.0017 |          0.0041 |                                       1 |
 | Allergy |          -0.0008 |         -0.0001 |                                       4 |
 | CAD   |          -0.0018 |         -0.0009 |                                       0 |
 | RA     |           0.0021 |          0.0037 |                                       1 |

The next step is to run fcFDR iteratively on the IMDs.

### What about the principal trait Z-score transformation?

I earlier transformed the auxiliary p-values with `-qnorm(./2)` as this is the transformation applied to the principal p-values in `fcfdr::flexible_cfdr`. `-qnorm(./2)` induces a negative correlation between the principal p-values and the Z-scores. We switched to `qnorm(./2)` to ensure a positive correlation between auxiliary p-values and Z-scores.

However, in the scenario that the principal and auxiliary p-values are positively correlated, i.e. the scenario of 'enrichment', using these divergent transformations will induce a negative correlation between the principal and auxiliary Z-scores: a pair of p-values `(p = 0, q = 0)` would be mapped to `(Inf, -Inf)`. Is this really what we want? Note that the auxiliary Z-scores will still be positively correlated with the principal p-values.

By default, `fcfdr::flexible_cfdr` enforces a positive Spearman's rank correlation in the independent SNP subset between the principal p-values and the auxiliary values by flipping the sign on the latter (in this case, the auxiliary Z-scores; the procedure never actually sees the auxiliary p-values). This is to ensure positive stochastic monotonicity (low p is enriched for low q). My experience thus far, however, is that the value of this correlation tends to be very small and can take both positive and negative values for data sets in which enrichment is equally plausible: for example, both T1D and the allergy ('aster') data sets above yield four novel signals, but the sign of Spearman's rank differs. The results I have obtained so far have had this enforcement turned off.

I don't have any answers to this yet. 

### LDAK revisited

LDAK requires a set of SNPs as input. At present, we join the GWAS SNPs to the hg19 1000G SNPs using rsIDs, then pass the intersection to LDAK. An alternative is to disregard rsIDs and instead join using chr:bp coordinates and allele pairs. I have tried both joining approaches in the past and went with rsIDs for convenience. But does one approach yield more SNPs in the join than the other? And does using GRCh38/hg38 1000G data change the number of SNPs retained as well?

NB: The GWAS-1000G joins below are performed after first joining each IMD data set to the PID data.

| Data set | # SNPs, rsID hg19 | # SNPs, chr/bp GRCh38 | # SNPs, chr/bp hg19 |
|----------|-------------------|-----------------------|---------------------|
| Allergy  | 5,804,169         | 5,771,438             | 5,843,679           |
| Asthma   | 5,475,376         | 5,409,116             | 5,475,968           |
| CD       | 5,785,364         | 5,748,211             | 5,803,317           |
| RA       | 4,464,034         | 5,605,512             | 5,680,329           |
| SLE      | 5,592,031         | 5,547,333             | 5,609,725           |
| T1D      | 5,596,884         | 5,552,579             | 5,649,090           |
| UC       | 5,788,435         | 5,751,271             | 5,806,505           |

In each case, joining on chr/bp with hg19 comes out on top. I could not locate GRCh38 1000G data which featured rsIDs, but when joining on chr/bp I could port over the rsIDs from the GWAS data. 

Speaking of joins...

### How many SNPs do we lose in each join?

The workflow at the moment is to join an auxiliary GWAS data set to:

* the PID data set, then
* the T1D MAF data, then
* the 1000G data

and as one might expect there is some SNP attrition at each step. Note that, at present, we join to the T1D MAF data, then to the LDAK data.

| Data set | No. in GWAS file | No. in PID join | No. in MAF join | No. LDAK join |
|----------|------------------|-----------------|-----------------|---------------|
| pid      | 7,290,874        | n/a             | n/a             | n/a           |
| ra       | 8,747,962        | 6,109,631       | 4,242,819       | 4,224,318     |
| t1d      | 9,037,957        | 5,668,167       | 5,661,097       | 5,596,884     |
| sle      | 7,914,824        | 5,760,149       | 5,255,630       | 5,237,182     |
| uc       | 9,486,539        | 6,540,096       | 5,333,396       | 5,320,462     |
| cd       | 9,469,592        | 6,531,342       | 5,332,457       | 5,319,543     |
| aster    | 13,794,598       | 6,581,058       | 5,469,110       | 5,413,955     |
| asthma   | 7,488,535        | 5,588,691       | 5,087,326       | 5,085,381     |

(Not sure how I lost ~7,000 values from the T1D-PID join to the MAF join with the same data set)

How many SNPs do we lose at each step?

| Data set | No. lost in PID join | No. lost in MAF join | No. lost LDAK join |
|----------|----------------------|----------------------|--------------------|
| ra       | 2,638,331            | 1,866,812            | 18,501             |
| t1d      | 3,369,790            | 7,070                | 64,213             |
| sle      | 2,154,675            | 504,519              | 18,448             |
| uc       | 2,946,443            | 1,206,700            | 12,934             |
| cd       | 2,938,250            | 1,198,885            | 12,914             |
| aster    | 7,213,540            | 1,111,948            | 55,155             |
| asthma   | 1,899,844            | 501,365              | 1,945              |

SNPs lost as a percentage of the SNPs in the original GWAS data set

| Data set | % lost in PID join | % lost in MAF join | % lost LDAK join |
|----------|--------------------|--------------------|------------------|
| ra       |              30.16 |              21.34 |             0.21 |
| t1d      |              37.28 |               0.08 |             0.71 |
| sle      |              27.22 |               6.37 |             0.23 |
| uc       |              31.06 |              12.72 |             0.14 |
| cd       |              31.03 |              12.66 |             0.14 |
| aster    |              52.29 |               8.06 |             0.40 |
| asthma   |              25.37 |               6.70 |             0.03 |

Can we save more SNPs? Some ideas:

* calculate MAF from the 1000G data in lieu of joining to the T1D data set
* treat as dependent those SNPs not present in the 1000G data, i.e. do not include them in KDE fitting, but do not discard them
