# May 12th, 2021

## TODOs

1. Work up the PID-four IMD (including IBD) fcFDR results (in progress, results so far below)
2. Redo GPS test with censored PID p-values [0, 0.01] and select auxiliary traits (done)


## What I've done this week

### GPS test statistics

We discussed this on Slack, but to recap: I added IgAD and IBD to the set of traits tested. IgAD is not related to the set of IMDs considered by Thaventhiran et al. so I did not consider it for inclusion, whereas IBD combines UC and CD, which were in the IMD set. aster, t1d, ibd, and ra came in the top four (actually cd did, but I will consider that ibd subsumes it for the time being) for both the whole set and [0, 0.01] subset, but their order differed (aster, t1d, ibd, and ra, and aster, ibd, t1d, and ra, respectively). I decided to go with the order from the whole set results, as that is the intended use case of the GPS test statistic.

| aux_traits |    nrow |        gps | threshold |        pval |
|------------|---------|------------|-----------|-------------|
| aster      | 1099563 | 15.2428479 |     1e+00 | 0.004294707 |
| igad       | 1140949 |  12.696673 |     1e+00 | 0.006184062 |
| t1d        | 1197808 |  9.9052761 |     1e+00 | 0.010140410 |
| ibd        | 1160475 |   6.853268 |     1e+00 | 0.021066353 |
| ra         | 1022495 |  6.6377017 |     1e+00 | 0.022441161 |
| cd         | 1159486 |  6.3361312 |     1e+00 | 0.024601073 |
| uc         | 1160298 |  4.9944806 |     1e+00 | 0.039295545 |
| asthma     | 1028369 |  3.5648396 |     1e+00 | 0.075673655 |
| sle        | 1009863 |  2.6500769 |     1e+00 | 0.132718059 |
| t2d        | 1188033 |  1.6420617 |     1e+00 | 0.309865973 |
| cad        | 1116740 |  1.2689268 |     1e+00 | 0.462620358 |
|------------|---------|------------|-----------|-------------|
| aster      |   11414 |  7.8020465 |     1e-02 | 0.016293731 |
| ibd        |   11866 |   6.828832 |     1e-02 | 0.021215777 |
| cd         |   11865 |  5.2813481 |     1e-02 | 0.035216693 |
| t1d        |   12257 |  4.4518711 |     1e-02 | 0.049204460 |
| ra         |   10649 |  3.8339127 |     1e-02 | 0.065769764 |
| uc         |   11902 |  3.5494438 |     1e-02 | 0.076305792 |
| igad       |   11766 |   2.799092 |     1e-02 | 0.119824362 |
| asthma     |   10775 |  1.9501241 |     1e-02 | 0.231220742 |
| t2d        |   12213 |  1.8319754 |     1e-02 | 0.257670376 |
| sle        |   10626 |  1.7036874 |     1e-02 | 0.291442844 |
| cad        |   11538 |  0.8123088 |     1e-02 | 0.780303050 |
|------------|---------|------------|-----------|-------------|

### PID-four IMD fcFDR

The following depicts final v-values output by the procedure for each set of four IMDs:

![](/images/120521/ibd_vs_non_ibd.png)

![](/images/120521/pid_four_imd_small.png)

I spent time thinking about how to automate the workup of our genome-wide significant SNPs and worked on code to fetch annotations (this took up most of my time). To identify lead SNPs, on each chromosome I sorted by v-value, picked the most significant SNP as the 'lead SNP', then excluded SNPs in a window of 500kbp and iterated. 

[GWS loci in fcFDR](/entries/120521/gws_loci.html)
