# 3rd February 2021

## TODOs from last meeting

1. Work up IgAD Bronson et al. data set
2. Run LDAK to select SNPs for Gaussian mixture fit
3. Write on PID
4. Fix icFDR script

## What I've done this week

### Reading and writing

Section on PID due tomorrow.

I will not be writing a section this coming week, as I plan to present a paper at the group meeting ('Droplet scRNA-seq is not zero-inflated').

### icFDR 

We talked last week about how the Gaussian mixture fit to the asthma p-values to estimate the distribution of `Q|H_P=0` differed between:
- the intersection of PID and asthma SNPs ('pairwise cFDR')
- the intersection of PID SNPs and those from five immune-mediated diseases: asthma, T1D, SLE, UC, CD

In both cases we fitted the mixture to asthma p-values. 

	> cfdr_fit
	$pars
	[1] 0.9125418 1.5625372
	> icfdr_fit
	$pars
	[1] 0.9974126 6.0104884

The difference was appreciable, but is there an obvious difference in the empirical distribution of the auxiliary trait p-values in the set `q[p > 0.5]`? Apparently not, at least not as can be shown by these plots.

![](/images/030221/qValHist_pairwiseVsFiveIMDs.png)

I am working on running LDAK to identify an independent set of SNPs to which to fit the mixture to see if that resolves the issue. Guille has been helping me with this.

### IgAD GWAS (Bronson et al. 2016)

I have done some initial work on this data set.

![]('/images/030221/pidVsIgadQQ.png')

The largest -log10(p) value for IgAD (sans MHC) is 35.83 at rs116540075, chr6:bp33,502,882 in hg38 (note that the MHC in hg38 is located at chr6:28,510,120-33,480,577). So this SNP is barely outside of the MHC, just over 22kbp away. After that peak, the next is rs1990760 at chr3:162,267,541 with 14.43. There's one nearby with 13.48. If we threshold at -log10(p)<13 (as in the Manhattan plot below), we only chop these two off (besides those on chr6).

![]('/images/030221/pidVsIgadHist.png')

I produced Manhattans for PID, IgAD, and the cFDR results. 

![]('/images/030221/igad_pid_cfdr.png')

I did the same using only the SNPs in the intersection of PID and IgAD (i.e. those for which we produce a v-value) for all three plots, too.

![]('/images/030221/igad_pid_cfdr_intersection.png')

