# 5th May, 2021

## TODOs 

1. Enrichment metrics
2. FDR regression
3. Work-up of fcFDR results

## What I've done

### fcFDR and FDR regression

No noteworthy progress on these as I spent most of my time reading, both about PID and GWAS in general.

### Enrichment metrics

#### Quantile/AUC metric 

Recall your definition:

```
y_alpha = | { i : p_i <= p_alpha, q_i <= q_alpha } | / | { i : q_i <= q_alpha } |
```

Last time we discussed this, I was using as a measurement of enrichment an estimate of the conditional probability `P( P < p_alpha | Q < q_alpha)`, where `P` and `Q` are the random variables from which are drawn p-values of association for the principal and auxiliary traits, respectively. This estimate `y_alpha` was then divided by alpha, the quantile. We see enrichment when `y_alpha / alpha > 1` and 'depletion' when `y_alpha / alpha < 1`. 

I found this metric was sensitive to the choice of quantile at which it was evaluated and Chris suggested I consider instead the area under the conditional probability estimate `y_alpha` as a cumulative measure. This can be divided by the expected area, that of a triangle, to test for enrichment using the relative AUC.

The table below shows the cumulative, relative AUC up to each quantile. Enrichment falls off for less significant p-values, hence the measure tending towards 1 at higher quantiles. The results are neater than those obtained with the previous approach. The `perm` column was obtained by permuting the order of `aster` values to produce an artificial null. `cad` and `t2d` are included as organic controls. I excluded the MHC for all data sets.

| Quantiles |  aster |   uc |   cd |   sle |    t1d | asthma |    ra |  cad |  t2d | perm |
|-----------|--------|------|------|-------|--------|--------|-------|------|------|------|
|     1e-05 |   0.00 | 0.00 | 0.00 |  0.00 |   0.00 |   0.00 |  0.00 | 0.00 | 0.00 | 0.00 |
|     5e-05 | 981.60 | 0.00 | 0.00 |  0.00 |   0.00 |   0.00 |  0.00 | 0.00 | 0.00 | 0.00 |
|     1e-04 | 959.21 | 0.00 | 0.00 |  0.00 | 187.17 |   0.00 |  0.00 | 0.00 | 0.00 | 0.00 |
|     5e-04 | 213.36 | 0.00 | 7.42 | 32.55 | 140.42 |  27.77 | 23.32 | 0.00 | 0.00 | 1.47 |
|     1e-03 |  75.92 | 0.69 | 7.42 | 25.59 |  71.22 |  24.66 | 19.60 | 0.00 | 0.29 | 1.37 |
|     5e-03 |   8.50 | 1.21 | 2.29 |  4.94 |   9.52 |   5.88 |  5.15 | 1.25 | 0.97 | 1.17 |
|     1e-02 |   4.13 | 1.20 | 1.37 |  3.93 |   5.56 |   2.96 |  2.84 | 1.36 | 1.02 | 1.11 |
|     5e-02 |   1.48 | 1.03 | 1.13 |  1.81 |   1.99 |   1.21 |  1.26 | 1.08 | 0.99 | 1.02 |
|     1e-01 |   1.19 | 1.01 | 1.09 |  1.28 |   1.34 |   1.05 |  1.09 | 1.03 | 1.00 | 1.01 |
|     5e-01 |   1.01 | 1.00 | 1.01 |  1.02 |   1.03 |   1.00 |  1.00 | 1.01 | 1.00 | 1.00 |
|     1e+00 |   1.00 | 1.00 | 1.00 |  1.01 |   1.01 |   1.00 |  1.00 | 1.00 | 1.00 |  1.0 |

Selection of quantiles at which to evaluate this measure is a consideration. The extreme approach would be to fit the ecdf, i.e. measure it at all (empirical) quantiles. As it happens, the next measure I considered does just that.

#### GPS test

This test was introduced by Li et al. in their 2015 paper, a meta-analysis of ten paediatric age-of-onset autoimmune diseases including CVID. It is intended as a test of the genome-wide pairwise association signal (hence 'GPS'). It is a genome-wide test as there is no restriction to particular loci, such as the genome-wide significant ones. 

![](/images/050521/gps_stat.png)

u and v denote a pair of p-values for the same SNP but association with different traits. We assume i.i.d. u and v, so I used the LDAK weights to pick out independent SNPs. 

I evaluated the test statistic for all pairs then took the maximum, but on reflection perhaps `sup` means I am meant to maximise this function over `[0,1] x [0, 1]`? As it happens, implementing this test entailed writing a parallelised bivariate ecdf in `Rcpp`, for which I had to learn basic use of `RcppParallel`. Even with that, it still takes ~20 minutes (with an admittedly naive ecdf implementation) to compute the statistic. Maximising the function numerically would probably be completely impractical. 

Li et al. claim the test statistic's distribution under the null approximates the inverse square root of a standard exponential random variable, so we transform its value accordingly before applying `pexp`. 

| Trait  | # SNPs    | GPS stat. | p-value |
|--------|-----------|-----------|---------|
| aster  | 1,099,563 | 15.242848 |  0.0043 |
| asthma | 1,028,369 |  3.564840 |   0.076 |
| cd     | 1,159,486 |  6.336131 |   0.025 |
| ra     | 1,022,495 |  6.637702 |   0.022 |
| sle    | 1,009,863 |  2.650077 |    0.13 |
| t1d    | 1,197,808 |  9.905276 |   0.010 |
| uc     | 1,160,298 |  4.994481 |   0.039 |
| cad    | 1,116,740 |  1.268927 |    0.46 |
| t2d    | 1,188,033 |  1.642062 |    0.31 |

#### Comparing the metrics

| Trait  | AUC rank | GPS stat. rank | Newly significant signals in fcFDR |
|--------|----------|----------------|------------------------------------|
| aster  |        1 |              1 |                                  4 |
| asthma |        4 |              6 |                                  0 |
| cd     |        6 |              4 |                                  0 |
| ra     |        5 |              3 |                                  1 |
| sle    |        3 |              7 |                                  1 |
| t1d    |        2 |              2 |                                  3 |
| uc     |        7 |              5 |                                  1 |
| cad    |        8 |              9 |                                  0 |
| t2d    |        9 |              8 |                                n/a |

Not as much concordance as I would like to see here, but the two are at least consistent for the two traits with the strongest signal; these two, `aster` and `t1d`, also yielded the most newly significant signals in pairwise fcFDR analyses (these are small numbers, though). If we take only the IMDs, Spearman's rank correlation is 0.43 for the two metrics.

### Another look at the iterative fcFDR results

![](/images/050521/pid_seven_imd.png)

![](/images/050521/icfdr_v_uc_it_7.png)
