# 10th February 2021

## TODOs from last meeting

1. Work up IgAD Bronson et al. data set
2. Run LDAK to select SNPs for Gaussian mixture fit
3. Write on PID
4. Fix icFDR script

## What I've done this week

### icFDR 

#### Differences in fitted mixture parameters are an artifact of my code

When writing the code to superimpose density estimates on the histograms we looked at last week, I thought to compare the estimates from `cfdr::fit.2g` and `cfdr::fit.2g.parallel`, my parallel implementation, as I have had trouble with the latter in my `icfdr` script. I have tested that the two implementations produce very similar estimates in the past and wrote unit tests to that end. 

The `cfdrQ` fits use the PID-asthma join asthma p-values and the `icfdrQ` use the PID-five IMD asthma p-values (a subset of the former), both thresholded to retain only those for which the corresponding PID p-value is greater than 0.5:

    cfdrQ_mixturePars
    [1] 0.9125358 1.5625140
    cfdrQ_mixturePars.parallel
    [1] 0.9125418 1.5625372
    icfdrQ_mixturePars
    [1] 0.9042702 1.5494240
    icfdrQ_mixturePars.parallel
    [1] 0.9974126 6.0104884
  
So the fits are similar for the `cfdrQ` data, but dissimilar for the `icfdrQ`. It seems the outrageous parameter estimates I saw for `icfdrQ` were due to my parallel implementation producing results divergent from those from James's serial implementation. The sole change made to James's `fit.2g` code was to replace the call to `optim` with `optimParallel::optimParallel`. Further, the `icfdrQ` data set was still large, (2,315,224 vs. 2,764,055 in the `cfdrQ` data set), so I do not think this is merely the result of a significantly smaller `icfdrQ` data set exaggerating the differences between the two implementations. 

Running LDAK was not a waste of time, though, as I now know how to do it if I use `fcfdr` in the future (probable) and if I need to generate a set of independent SNPs for another fitting purpose. I would like to see how the estimates differ when the mixture is fit to the independent SNPs, so I'm going to go ahead with that.

#### Update: how do the log-likelihoods differ between the estimates?

Both `fit.2g` and `fit.2g.parallel` use `L-BFGS-B`. `h0value` gives the log-likelihood of the standard normal component and `h1value` the log-likelihood under the `N(0, sigma^2)` component.

    cfdrQ_mixture
    $pars
    [1] 0.9125358 1.5625140
    $h1value
    [1] -4099639
    $h0value
    [1] -4166801

    cfdrQ_mixture.parallel
    $pars
    [1] 0.9125418 1.5625372
    $h1value
    [1] -4099639
    $h0value
    [1] -4166801

    icfdrQ_mixture
    $pars
    [1] 0.9042702 1.5494240
    $h1value
    [1] -3440721
    $h0value
    [1] -3495125

    icfdrQ_mixture.parallel
    $pars
    [1] 0.9974126 6.0104884
    $h1value
    [1] -3448889
    $h0value
    [1] -3495125
  
I'm not sure whether such large negative values are normal even with log-likelihoods, but the parallel fit to the `icfdrQ` values attains a slightly smaller log-likelihood with its odd parameters. 

### IgAD GWAS (Bronson et al. 2016)

I took a closer look at the top extra-MHC SNP rs116540075. Although it does not fall into the MHC interval for hg38 (chr6:28,510,120-33,480,577), the plot below shows the boundary in terms of the p-value behaviour is not discrete. I would put its high p-value (and those of its neighbours) down to its being in LD with something inside the nominal MHC; this is best seen in the first plot. Note the 0-137 axis on a negative log scale; the solid grey line near the bottom is genome-wide significance.

![](/images/100221/igadZoom.png)

dbSNP turns up nothing on this SNP and it does not appear to be located in any sort of annotated genomic feature as judged by looking at the locus in the UCSC Genome Browser.

![](/images/100221/igadZoomToTopSnp.png)

In future, perhaps it would be best for me to identify the MHC empirically by looking at a plot like the first one?

