# 17th March 2021

## TODOs from last meeting

1. Implement sampling scheme discussed
2. Write about PID and autoimmunity
3. Query about the SLE peaks

## What I've done

### Matching the MAF distribution of the whole SNP set in the ind. subset

We want to downsample the independent SNPs so that their distribution matches that of the whole set of SNPs. In the matching scheme you suggested, the size `n` of the largest (down)sample we can draw depends on the minimum of  `absoluteFreq.ind / relativeFreq.whole` across all bins. For a given bin we can't sample more independent SNPs than there are in the bin to begin with: `n * relativeFreq.whole <= absoluteFreq.ind`.

For my MAF bins I selected breaks `(0, 0.01], (0.01, 0.02], ..., (0.49, 0.5]`. Another approach to bin width selection would be to use the Freedman-Draconis rule (`2*IQR/n^(1/3)`): this gives breaks `(0,0.02], (0.02, 0.04], ...`. There was little difference in the maximum sample size I got with each, so I went with the first.

I was able to draw a sample of ~427k independent SNPs with the matched distribution using a T1D data set (chosen as it is the only one with MAF data to hand). Compare this figure with the 1.2M independent SNPs (and 4.35M dependent SNPs) in the original data set. 

I looked at the density of SNPs over the genome in both the independent subset and the downsampled set.

![](/images/170321/t1d_matched_maf_independentSNPs_only_density_chr1_11.png)
![](/images/170321/t1d_matched_maf_independentSNPs_only_density_chr12_22.png)

The gross outlook is that coverage doesn't seem to differ in any systematic way.

(I noticed the lack of density over the p-arm of chromosomes 13, 14, 15, 21, and 22, and thought something was awry, but then into my head popped a word I hadn't heard since about 2013 or 2014: 'acrocentric'.)

We expect the shape of the downsampled subset histogram to match that of the whole set, and it does:

![](/images/170321/downsampling_histograms.png)

Lastly, the results of `fcfdr` run using the matched independent subset. 

![](/images/170321/t1d_matched_maf.png)

### Query about SLE peaks from last week

You asked if a particular peak on chromosome 6 in the PID-conditioned-on-SLE results was the shared between the `cfdr` and `fcfdr` or whether it was distinct. First the image from last week for reference:

![](/images/100321/sle_pid_backToBack_threeTrack.png)

Upon closer inspection, the peaks are not distinct:   
 
![](/images/100321/sle_pid_backToBack_chr2_6_16.png)

### Writing on PID and autoimmunity

I will have this ready for Wednesday morning.

### Gaussian mixture fits using downsampled ind. SNPs

Without the use of weights:

```
est_q0_ind <- fit.2g(innerJoinDat[P.2 > 0.5 & Ind == T]$P.1)
$pars
[1] 0.8580436 1.4024762
est_q0_ind_downsample <- fit.2g(innerJoinDat[ind_snps][P.2 > 0.5]$P.1)
[1] 0.8626373 1.4586058
```

If I pass weights to `cfdr::fit.2g`, the estimate sticks at the initial values `(0.5, 1.5)`. 

