# 3rd March 2021

## TODOs from last meeting

(I don't list all of my weekly todos here, just the ones arising from our previous meeting)

- create plots of the `fcfdr` results we do have
- investigate distribution of SNPs by MAF in data sets; what does the distribution of rare SNPs look like?
- investigate relationship between MAF and p-values
- how does MAF affect likelihood of a SNP's being retained by LDAK?

## What I've done

### Comparative Manhattan plots 

First, a question for tomorrow: how impressed should I be with a lone SNP (or a pair or a trio) which is highly significant? Perhaps this is how those rare, independent SNPs crop up as we discussed last week?

Second, a point I want to come back to (somewhat related to the first). The joining of a set of SNPs from the principal data set to the auxiliary data set(s) means we lose SNPs and this is compounded by joining to the 1000G data set in the case of `fcfdr`. I think this means we should be more forgiving of sparse signals.

For the four data sets on which I was able to run `fcfdr`, I created 'back-to-back' Manhattan plots comparing the results obtained with `fcfdr` and `cfdr`. In both cases, you are looking at the v-values produced by conditioning PID p-values on p-values from the named IMD. For reference, here's the PID Manhattan:

![](/images/030321/pid.png)

Thaventhiran et al. did GWAS of antibody-deficient PID and 'all PID', and meta-analysed their results with those from Li et al. (a 2015 CVID GWAS). I believe the p-values I am using are those from the 'marginal' antibody-deficient PID GWAS, but I'm not 100% sure.

(I don't expect you to look at all of these, it was just easy for me to look at them by putting them on the web)

[PID conditioned on SLE](/entries/030321/sle_pid.html)

[PID conditioned on UC](/entries/030321/uc_pid.html)

[PID conditioned on CD](/entries/030321/cd_pid.html)

[PID conditioned on T1D](/entries/030321/t1d_pid.html)

Two things to consider:
* `cfdr` and `fcfdr` vs. the original GWAS
* `cfdr` vs. `fcfdr` 

Summarising by auxiliary trait:
* SLE: Newly genome-wide significant (GWS) loci, STAT4 and CLEC16A, and a near-miss at IRF8 with `cfdr`, nothing with `fcfdr`
* UC: PTPN2 signal reached GWS with `cfdr`, nothing with `fcfdr`
* CD: PTPN2 signal reached GWS with `cfdr`, nothing with `fcfdr` 
* T1D: PTPN22 [sic], CLEC16A, and PTPN2/LINC01882 reached GWS with `cfdr`, nothing with `fcfdr`

I don't believe the PTPN22 signal was mentioned in the original paper.

The fact that I got no GWS hits with `fcfdr` makes me think that I am losing signals through filtering only for those SNPs present in the 1000G data (this for the sake of the LDAK weights). The GWS signals listed for `cfdr` are sometimes only one or two SNPs, so it's easy to imagine these are lost in the `fcfdr` filtering/joining. I am going to take a look at this ASAP.

### MAF and p-values

I had thought I had MAF information to hand, but it turns out to be missing from most of the IMD data sets we are using. Guille has shown me a script to calculate MAF using the 1000G data and a designated population, but I have some reservations about the accuracy of such estimated MAF after comparing the numbers to those on dbSNP, and about the choice of the population. On the latter point, so far I have gone with the Utah central European cohort ('CEU') on Guille's recommendation. 

The T1D GWAS data did contain MAF and a brief investigation has shown there is something going on. In particular, it seems you were right that rare SNPs are more likely to be independent. A lot more to come on this front.

![](/images/030321/t1dMAF_wholeAndIndSNPs.png)
